lib/validation.js

```diff
@@ -46,27 +46,27 @@ function build (opts) {
       }
     }
 
-    opts[querystringSchema] = ajv.compile(opts.schema.querystring)
+    opts[querystringSchema] = compile(opts.schema.querystring)
   }
 
   if (opts.schema.params) {
-    opts[paramsSchema] = ajv.compile(opts.schema.params)
+    opts[paramsSchema] = compile(opts.schema.params)
   }
 }
 
-function validate (handle, params, body, query) {
-  if (handle[paramsSchema] && !handle[paramsSchema](params)) {
-    return inputSchemaError(handle[paramsSchema].errors)
-  }
-
-  if (handle[bodySchema] && !handle[bodySchema](body)) {
-    return inputSchemaError(handle[bodySchema].errors)
-  }
+function validateParam (validatorFunction, request, paramName) {
+  var ret = validatorFunction && validatorFunction(request[paramName])
+  if (ret === false) return inputSchemaError(validatorFunction.errors)
+  if (ret && ret.error) return ret.error
+  if (ret && ret.value) request[paramName] = ret.value
+  return false
+}
 
-  if (handle[querystringSchema] && !handle[querystringSchema](query)) {
-    return inputSchemaError(handle[querystringSchema].errors)
-  }
-  return true
+function validate (store, request) {
+  return validateParam(store[paramsSchema], request, 'params') ||
+    validateParam(store[bodySchema], request, 'body') ||
+    validateParam(store[querystringSchema], request, 'query') ||
```

Comment:
We are using `querystring ` and not `query`:)

https://github.com/fastify/fastify/blob/master/docs/Validation-And-Serialize.md#validation
https://github.com/fastify/fastify/blob/master/test/internals/validation.test.js#L72-L87

Reply:
Ops, my bad, here you are accessing the `request` object :D

Reply:
Maybe we need to add all kind validation test